### ISSUE TITLE
Avoid imported protected attributes or functions from third-party packages

### ISSUE THREAD

**Issue by franknoe** [markovmodel/PyEMMA#615]
In order to avoid problems like #613 which will occur by upgrades of third-party packages, we should avoid importing any protected attributes or functions. These can be changed in future versions and will then break our code.

Please check the code for such occurances, especially mdtraj imports. One example is given in`pyemma.coordinates.util.patches`:

``` python
from mdtraj.core.trajectory import load, _parse_topology, _TOPOLOGY_EXTS, _get_extension, open
```

If we need things to be made part of the public API, we can request this by issue or PR to mdtraj. If this is not desired, we have to simply copy the protected code.

**Comment by marscher** [comment]
fixed by #614

**Comment by franknoe** [comment]
No, this isn't fixed. This refers to protected imports that are still there after #614

**Comment by marscher** [comment]
Which occurrences of non-public api usages are you talking about? At least in the featurizer there are no further usages after #614.

**Comment by franknoe** [comment]
Not in the featurizer. Please read the description of this issue.

**Comment by marscher** [comment]
right, we can write a little import hook script (a function is called with the name of every imported module/package or name), which checks for private names. But I suppose we have used them rarely, because it is a very bad practice.

**Comment by franknoe** [comment]
yes. I think we just need to check the mdtraj imports for now. I would 
be suprised if we do that elsewhere.

Am 09/11/15 um 00:59 schrieb Martin K. Scherer:

> right, we can write a little import hook script (a function is called 
> with the name of every imported module/package or name), which checks 
> for private names. But I suppose we have used them rarely, because it 
> is a very bad practice.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/615#issuecomment-154887907.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by rmcgibbo** [comment]
If there are other things that you need from mdtraj that aren't currently exported in a public API, I'm happy to add them to a new public API. We've got the release-engineering stuff automated enough now that it's very easy to cut new releases, and we're looking increase the release frequency going forward.

**Comment by franknoe** [comment]
thanks!

Am 09/11/15 um 01:04 schrieb Robert T. McGibbon:

> If there are other things that you need from mdtraj that aren't 
> currently exported in a public API, i'm happy to add them to a new 
> public API. We've got the release-engineering stuff automated enough 
> now that it's very easy to cut new releases, and we're looking 
> increase the release frequency going forward.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/615#issuecomment-154888179.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by marscher** [comment]
Thank you @rmcgibbo :+1:

**Comment by marscher** [comment]
@rmcgibbo Did you observed these error messages of conda recently?

https://ci.appveyor.com/project/marscher/pyemma/build/job/6ohp9ro39m1a2wrv

**Comment by rmcgibbo** [comment]
No, that one is new to me. I've seen the matplotlib libpng reported here: https://github.com/conda/conda/issues/1753. Did you file a bug upstream for that one?

**Comment by marscher** [comment]
There are distributed issues more or less pointing to the same issue:
https://github.com/conda/conda/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+assertionerror+matchspec

One of the main conda devs is on vacation and the repo has some inconsistencies, which have to be figured out by fairly new people I guess.

**Comment by rmcgibbo** [comment]
Got it. Also, Aaron [left for a new job](http://asmeurer.github.io/blog/posts/lessons-learned-from-working-at-continuum/). Windows packaging is also just really _hard_. I don't begrudge them at all.

**Comment by gph82** [comment]
Hi, so I guess we can copy (and mantain?) the topology extension list ( _TOPOLOGY_EXTS) no problem.

What about _parse_topology and  _get_extension (https://github.com/mdtraj/mdtraj/blob/master/mdtraj/core/trajectory.py#L149)? 

Copying or into the new public API? I don't really have an opinion there (don't know enough about best practices in design etc)

**Comment by gph82** [comment]
BTW, same problem with _FormatRegistry here
https://github.com/markovmodel/PyEMMA/blob/b23e1d3e350bbde6680df21463d732b813472a6c/pyemma/coordinates/data/traj_info_cache.py#L36
and here:
https://github.com/markovmodel/PyEMMA/blob/21bbe600d044a2e0e59b280dba6e2510ba2a5158/pyemma/coordinates/data/util/reader_utils.py#L81

![snap6](https://cloud.githubusercontent.com/assets/7518004/11031907/1f8e198e-86d8-11e5-9510-6e66d5fd7e0b.png)

**Comment by marscher** [comment]
A public api, which file extension can be read, would be nice to have.

**Comment by franknoe** [comment]
I agree with Martin: It would be useful to have that exported as a 
public API, because it's on mdtraj to define which formats they can / 
wanna support.

Am 09/11/15 um 11:38 schrieb Guillermo Pérez-Hernández:

> Hi, so I guess we can copy (and mantain?) the topology extension list 
> ( _TOPOLOGY_EXTS) no problem.
> 
> What about _parse_topology and _get_extension 
> (https://github.com/mdtraj/mdtraj/blob/master/mdtraj/core/trajectory.py#L149)? 
> 
> Copying or into the new public API? I don't really have an opinion 
> there (don't know enough about best practices in design etc)
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/615#issuecomment-155024885.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by marscher** [comment]
mdtraj 1.6.1 exposes the FileRegistry now.
